gpio.py

```diff
@@ -68,17 +68,35 @@ def wrapped(pin, *args, **kwargs):
     return wrapped
 
 
-def cleanup(pin):
+def cleanup(pin=None, assert_exists=False):
+    """Cleanup the pin by closing and unexporting it.
+
+    Args:
+        pin (int, optional): either the pin to clean up or None (default).
+            If None, clean up all pins.
+        assert_exists: if True, raise a ValueError if the pin was not
+            setup. Otherwise, this function is a NOOP.
+    """
+    if pin is None:
+        # Take a list of keys because we will be deleting from _open
+        for pin in list(_open):
+            cleanup(pin)
+        return
+    if not isinstance(pin, int):
+        raise TypeError("pin must be an int, got: {}".format(pin))
     if pin not in _open:
+        if assert_exists:
+            raise ValueError("pin {} was not setup".format(pin))
         return
-    files = _open[pin]
-    files['value'].close()
-    files['direction'].close()
-    if os.path.exists(gpiopath(pin)):
-        log.debug("Unexporting pin {0}".format(pin))
-        with _export_lock:
+    with _export_lock:

```

Comment:
I moved the `export_lock` down because I really don't want people opening/closing files while this is happening. On that note, we should probably just have a lock per pin to support multi-threaded operation better.